lib/vrl/tests/tests/expressions/if_statement/error_predicate.vrl

```diff
@@ -0,0 +1,16 @@
+# result:
+# error[E111]: fallible predicate
+#   ┌─ :2:4
+#   │
+# 2 │ if contains(.noog, "thing") {
+#   │    ^^^^^^^^^^^^^^^^^^^^^^^^
+#   │    │
+#   │    this predicate can result in runtime error
+#   │    handle the error case to ensure runtime success
+#   │
+#   = see documentation about error handling at https://errors.vrl.dev/#handling
+#   = see language documentation at https://vrl.dev
```

Comment:
Thinking this through in the context of the [improved diagnostics RFC](https://github.com/vectordotdev/vector/blob/5f8a3a76003bdf44bebd6c5e78bfb34d12313518/rfcs/2021-08-22-7204-vrl-error-diagnostic-improvements.md) that I'll be implementing soon. It occurs to me that this diagnostic message is similarly non-specific as the ["expression can result in runtime error"](https://github.com/vectordotdev/vector/blob/5f8a3a76003bdf44bebd6c5e78bfb34d12313518/rfcs/2021-08-22-7204-vrl-error-diagnostic-improvements.md#remove-generic-expression-can-result-in-runtime-error-diagnostic) case that I'll be working on to remove.

In this case, the compiler should fail to compile because `contains` becomes fallible due to `.noog` not being a string, so instead, it should've raised _that_ error.

If you see an easy fix for this that you can do in this PR, I'd love to see that happen, but if not, I'd love for you to create an issue to fix this, and link it to the diagnostics improvements RFC, so that I can pick it up when I get to it.

Reply:
This is far too involved a fix for this PR!

I think we would need to change fallibility from being a simple boolean to a more involved struct that tracks the underlying cause, including spans. Currently, by the time the fallibility reaches the predicate it has lost everything beyond the simple fact that something is fallible somewhere..

Reply:
Yeah, I was thinking more along the lines of allowing a "hey my expression is fallible because of a nested expression, so this can't compile, but I'm also not going to emit an error, my nested expression will" set-up that would prevent us from having to go so far as what you are suggesting.

But that's just an initial rough idea, and I'd need to validate if that can work first.

But yeah, I can see how that would be too involved to solve in this PR, I'll link it to #8894

Reply:
Ah, actually, I wrote down how I think we can solve this in https://github.com/vectordotdev/vector/issues/8925, which I suspect will also work in this case. We'll see once I get to resolving that RFC.